src/fklearn/causal/cate_learning/meta_learners.py

```diff
@@ -0,0 +1,231 @@
+        (df[treatment_col] == treatment_name) | (df[treatment_col] == control_name), :
+    ]
+    return treatment_control_df
+
+
+def _create_treatment_flag(
+    df: pd.DataFrame, treatment_col: str, treatment_name: str
+) -> pd.DataFrame:
+    df = df.copy()
+    treatment_flag = np.where(df[treatment_col] == treatment_name, 1.0, 0.0)
+    df[TREATMENT_FEATURE] = treatment_flag
+
+    return df
+
+
+def _fit_by_treatment(
+    df: pd.DataFrame,
+    learner: LearnerFnType,
+    treatment_col: str,
+    control_name: str,
+    treatments: list,
+) -> Tuple[dict, dict]:
+    fitted_learners = {}
+    learners_logs = {}
+    for treatment in treatments:
+        treatment_control_df = _filter_by_treatment(
+            df=df,
+            treatment_col=treatment_col,
+            treatment_name=treatment,
+            control_name=control_name,
+        )
+        treatment_control_df = _create_treatment_flag(
+            df=treatment_control_df,
+            treatment_col=treatment_col,
+            treatment_name=treatment,
+        )
+        learner_fcn, _, learner_log = learner(treatment_control_df)
+        fitted_learners[treatment] = learner_fcn
+        learners_logs[treatment] = learner_log
+
+    return fitted_learners, learners_logs
+
+
+def _predict_by_treatment_flag(
+    df: pd.DataFrame,
+    learner_fcn: PredictFnType,
+    is_treatment: bool,
+    prediction_column: str,
+) -> np.ndarray:
+
+    if is_treatment:
+        treatment_flag = np.ones(df.shape[0])
+    else:
+        treatment_flag = np.zeros(df.shape[0])
+
+    df[TREATMENT_FEATURE] = treatment_flag
+    prediction_df = learner_fcn(df)
+
+    return prediction_df[prediction_column].values
+
+
+def _simulate_treatment_effect(
+    df: pd.DataFrame, treatments: list, learners: dict, prediction_column: str
+) -> pd.DataFrame:
+    uplift_cols = []
+    scored_df = df.copy()
+    for treatment in treatments:
+        learner_fcn = learners[treatment]
+
+        scored_df[
+            f"treatment_{treatment}__{prediction_column}_on_treatment"
+        ] = _predict_by_treatment_flag(
+            df=scored_df,
+            learner_fcn=learner_fcn,
+            is_treatment=True,
+            prediction_column=prediction_column,
+        )
+        scored_df[
+            f"treatment_{treatment}__{prediction_column}_on_control"
+        ] = _predict_by_treatment_flag(
+            df=scored_df,
+            learner_fcn=learner_fcn,
+            is_treatment=False,
+            prediction_column=prediction_column,
+        )
+
+        uplift_cols.append(f"treatment_{treatment}__uplift")
+        scored_df[uplift_cols[-1]] = (
+            scored_df[f"treatment_{treatment}__{prediction_column}_on_treatment"]
+            - scored_df[f"treatment_{treatment}__{prediction_column}_on_control"]
+        )
+
+    scored_df["uplift"] = scored_df[uplift_cols].max(axis=1).values
+    scored_df["suggested_treatment"] = scored_df[uplift_cols].idxmax(axis=1).values
```

Comment:
we should also recommend not doing anything in the case the uplift is negative

Reply:
✅